### ISSUE TITLE
Hanging punctuation should include hyphens, particularly those used with justify [css-text-3] [css-text-4]

### ISSUE THREAD

**Issue by kevinmarks** [w3c/csswg-drafts#9329]
The [hanging-punctuation](https://www.w3.org/TR/css-text-3/#hanging-punctuation-property) property should include hyphens as well as stops and commas.

When `text-align: justify` or equivalent is active, setting `hanging-punctuation: force-end` should hang all hyphens, whether typed or inserted by the user agent. Ideally this would be on by default in this mode, to replicate good typographic practice with fully justified text.

This applies to [css-text-4](https://drafts.csswg.org/css-text-4/#hanging-punctuation-property) as well, and may be more relevant there with changes to line wrapping.

**Comment by fantasai** [comment]
The question I have here is:
- should it only be hyphens, or do people who want this want other punctuation to hang as well? Note that the current *-end values only hang CJK punctuation (on purpose).
- should the hyphens hang 100% or 50% or ...? I feel like I've seen various practices here.

**Comment by kizu** [comment]
I'll try to answer as someone who wanted to have hanging punctuation in web for years and years, with manual implementation of different kinds of success.

1. In my practice, I'd want to be able to hang: automatic hypens, `&shy;`, em/en-dash, quotation symbols, and probably more, like periods etc.
2. As an author, _ideally_ I'd want to be able to control the amount of hanging, potentially based on the symbol and side. And this can be dependent on the context: for headers we usually would want to hang things by 100%, for `text-align: justify` we could want to choose the value in a way it would be visually pleasing based on the font in question.

## [Reference 1](https://www.artlebedev.com/mandership/120/)

In the following screenshots we can see full hanging on one side, and partial on another for one symbol, and full for another.

<img width="873" alt="A screenshot of an example showing two blocks of text: one without hanging punctuation, another with it. The example contains punctuation on the left edge of the text, with the opening quotation and bracket hanging fully." src="https://github.com/w3c/csswg-drafts/assets/177485/47479566-80c9-47f2-bdf3-351bb452e8e1">
<img width="901" alt="A screenshot of an example showing two blocks of text: one without hanging punctuation, another with it. The example contains punctuation on the right edge of the text, with the hyphenation hanging by 50%, and a period hanging by 100%." src="https://github.com/w3c/csswg-drafts/assets/177485/7d5f860c-43eb-4c35-9066-0682996d7227">

## [Reference 2](https://indesignskills.com/tutorials/hanging-punctuation/)

Here the hanging punctuation is only on the left edge, and is different for all three cases: half for a quotation symbol inside the initial letter, very slight for an em-dash, and full for a quotation symbol in the body of text.

<img width="520" alt="A screenshot of an example from the article, showing a block of text with an initial letter and some guidelines showing the edges of the text." src="https://github.com/w3c/csswg-drafts/assets/177485/986fbfec-5426-4500-812f-0d3b2170fc03">

## [Reference 3](https://www.fonts.com/content/learning/fontology/level-4/fine-typography/hanging-punctuation) 

The first example on this page has 100% hanging punctuation on both sides, the second example is more interesting:

before | after
-|-
<img width="86" alt="A screenshot of a left edge of an example, showing paragraphs with the capital letters being indented, not aligning visually with the left edge of the text." src="https://github.com/w3c/csswg-drafts/assets/177485/c92a511c-b0c0-420d-8d9a-b436b72c566e"> | <img width="86" alt="A screenshot of a left edge of an example, showing the capital letters being adjusted to be optically aligned with the left edge of the text." src="https://github.com/w3c/csswg-drafts/assets/177485/6019b7b2-f2d9-4b21-9a50-7dbc0d9b2206">

Though, this is not about _punctuation_, this is something that I used to handle by assigning a custom className to the paragraph based on the first letter in it, and then adjusting the `::first-letter`. Not sure hanging-punctuation could help us with this case.

**Comment by fantasai** [comment]
My impression is that what we're really wanting here in most cases isn't hanging punctuation, it's optical alignment of the glyphs at the start/end of the line.

**Comment by kizu** [comment]
> My impression is that what we're really wanting here in most cases isn't hanging punctuation, it's optical alignment of the glyphs at the start/end of the line.

This is the case for the example in the third reference, but for the hyphens I found the 50% hanging to be usually preferable:

<img width="81" alt="Cropped screenshot of the example showing the right edge with the hyphens half-hanging over the text box's edge." src="https://github.com/w3c/csswg-drafts/assets/177485/00b1aa82-6600-4da6-a774-eb1a16d4beca">

While this _can_ be considered _optical_ alignment, I'd argue that any hanging punctuation can be treated as such if we squint hard enough.

Speaking of, as a fun sidenote anecdote, the way I did implement the hanging hyphens on web in the past was by modifying the font box of the hyphens in the font, adding `&shy;` wrapped in a span and assigning that modified font to these spans. A _very_ convoluted work-around, with alternatives, at the time, having some crossbrowser issues ([bug1](https://bugs.chromium.org/p/chromium/issues/detail?id=587918), [bug2](https://bugs.webkit.org/show_bug.cgi?id=153980)).

Having a native control over this would be really nice, as it would remove the necessity for these kinds of hacks.

**Comment by fantasai** [comment]
@kizu For optical alignment, my thought is that the best place to define it would be the font itself, like in a "start edge to X" pairs kerning table. And have a switch to enable that in CSS. (Idk if OpenType has such a thing... or even if it does, if any fonts implement it.) Because the amount by which you want to hang depends a lot on the design of the glyph.

> While this can be considered optical alignment, I'd argue that any hanging punctuation can be treated as such if we squint hard enough.

Kinda sorta, but there's a difference between "hang this character 100% outside the bounds" and "find an appropriate amount to hang it so it looks good". CJK mostly uses the latter, because it's gridded. But for Latin, we find it happening sometimes e.g. on pull quotes in particular, but that's usually a first/last thing rather than a start-edge/end-edge thing.

**Comment by kizu** [comment]
> For optical alignment, my thought is that the best place to define it would be the font itself, like in a "start edge to X" pairs kerning table. And have a switch to enable that in CSS. (Idk if OpenType has such a thing... or even if it does, if any fonts implement it.) Because the amount by which you want to hang depends a lot on the design of the glyph.

I completely agree that ideally it would be defined in the font per glyph. But I'm too far from understanding how fonts work inside, how hard would it be to propose/implement something like this as something available in the font editors etc, so I'm talking about how I would like to see it in CSS. It would be awesome if font authors could just provide this info, and web browsers could just use this info!

> But for Latin, we find it happening sometimes e.g. on pull quotes in particular, but that's usually a first/last thing rather than a start-edge/end-edge thing.

Usually, not not always: most cases in latin/cyrillic typography I saw were handling hanging cases in the middle of the text as well, not only in the beginning and end (I'm completely unfamiliar with CJK, so could only speak about latin/cyrillic cases).
